File: DesignPatternsInAutomatedTesting-Series/TemplateMethodDesignPattern/v0.1/Base/ShoppingCart.cs
// <copyright file="ShoppingCart.cs" company="Automate The Planet Ltd.">
// Copyright 2019 Automate The Planet Ltd.
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
// <author>Anton Angelov</author>
// <site>http://automatetheplanet.com/</site>

using TemplateMethodDesignPattern.Data.First;
using TemplateMethodDesignPattern.Pages.Checkout.First;
using TemplateMethodDesignPattern.Pages.Item.First;
using TemplateMethodDesignPattern.Pages.ShippingAddress.First;
using TemplateMethodDesignPattern.Pages.SignIn.First;

namespace TemplateMethodDesignPattern.Base.First
{
    public class ShoppingCart
    {
        private readonly IItemPage _itemPage;

        private readonly ISignInPage _signInPage;

        private readonly ICheckoutPage _checkoutPage;

        private readonly IShippingAddressPage _shippingAddressPage;

        public ShoppingCart(IItemPage itemPage, ISignInPage signInPage, ICheckoutPage checkoutPage, IShippingAddressPage shippingAddressPage)
        {
            _itemPage = itemPage;
            _signInPage = signInPage;
            _checkoutPage = checkoutPage;
            _shippingAddressPage = shippingAddressPage;
        }

        public void PurchaseItem(string item, double itemPrice, ClientInfo clientInfo)
        {
            _itemPage.Open(item);
            _itemPage.AssertPrice(itemPrice);
            _itemPage.ClickBuyNowButton();
            _signInPage.ClickContinueAsGuestButton();
            _shippingAddressPage.FillShippingInfo(clientInfo);
            _shippingAddressPage.AssertSubtotalAmount(itemPrice);
            _shippingAddressPage.ClickContinueButton();
            _checkoutPage.AssertSubtotal(itemPrice);
        }
    }
}

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/facade/ShoppingCart.java
package facade;

import facade.data.PurchaseInfo;
import facade.pages.interfaces.CheckoutPage;
import facade.pages.interfaces.ItemPage;
import facade.pages.interfaces.ShoppingCartPage;

public class ShoppingCart {
    private final ItemPage itemPage;
    private final ShoppingCartPage shoppingCartPage;
    private final CheckoutPage checkoutPage;

    public ShoppingCart(ItemPage itempage, ShoppingCartPage shoppingCartPage, CheckoutPage checkoutPage) {
        this.itemPage = itempage;
        this.shoppingCartPage = shoppingCartPage;
        this.checkoutPage = checkoutPage;
    }

    public void purchaseItem(String itemUrl, double itemPrice, PurchaseInfo purchaseInfo) {
        itemPage.navigate(itemUrl);
        itemPage.assertPrice(itemPrice);
        itemPage.clickBuyNowButton();
        itemPage.clickViewShoppingCartButton();
        shoppingCartPage.clickProceedToCheckoutButton();
        shoppingCartPage.assertSubtotalAmount(itemPrice);
        checkoutPage.fillBillingInfo(purchaseInfo);
        checkoutPage.assertSubtotal(itemPrice);
    }
}
